### DISCUSSION TITLE
Error force user reset password after restore with magiclink

### DISCUSSION THREAD

**Discussion by caosdp-rs** [codeigniter4/shield#644] [Q&A] [answered]
Hello!

I'm try force user change password after login with magiclink.

in Auth.php 

```php
public function loginRedirect(): string
    {
        if (session('magicLogin')) {
            return redirect()->route('set_password');
         }
```

`php spark routes -h `

`| GET     | change-password | » | \App\Controllers\PasswordReset::index | csrf session | toolbar |
`

but cause error 

Config\Auth::loginRedirect(): Return value must be of type string, CodeIgniter\HTTP\RedirectResponse returned

**Comment by kenjis** [comment, marked as answer]
> Config\Auth::loginRedirect(): Return value must be of type string, CodeIgniter\HTTP\RedirectResponse returned

`Config\Auth::loginRedirect()` must return URL string, not Response object.

**Reply by caosdp-rs** [reply]
[docs](https://github.com/codeigniter4/shield/tree/08bbc373e2598b630bcb39c10e9190d78873289b/docs)/quickstart.md


Session Notification
You can detect if a user has finished the magic link login by checking for a session value, magicLogin. If they have recently completed the flow, it will exist and have a value of true.

if (session('magicLogin')) {
    return redirect()->route('set_password');
}

**Reply by caosdp-rs** [reply]
this work!
```

  public array $redirects = [
        'register'    => '/',
        'login'       => '/manager',
        'logout'      => '/',
        'set_password'=> 'change-password-magic',
    ];
```


    public function loginRedirect(): string
    {
        if (session('magicLogin')) {
            $url = setting('Auth.redirects')['set_password'];
         }else{
            $url = setting('Auth.redirects')['login'];
         }
        return $this->getUrl($url);
    }

**Reply by kenjis** [reply]
The following code is a sample code for a controller.
```php
if (session('magicLogin')) {
    return redirect()->route('set_password');
}
```
https://github.com/codeigniter4/shield/blob/develop/docs/quickstart.md#session-notification

**Comment by datamweb** [comment]
Hi @caosdp-rs,
**Your method is good.**
```php
public function loginRedirect(): string
{
    if (session('magicLogin')) {
        $url = setting('Auth.redirects')['set_password'];
     }else{
        $url = setting('Auth.redirects')['login'];
     }
    return $this->getUrl($url);
}
```
But I saw something in your code that I would like to explain a little. You could write your code more readable and beautiful.
```php
public function loginRedirect(): string
{

 $url = setting('Auth.redirects')['login'];

  if (session('magicLogin')) {
      $url = setting('Auth.redirects')['set_password'];
   }

   return $this->getUrl($url);
}
```

As you can see, both of the above codes work correctly, but which one is better? 
[Write better code and be a better programmer by NEVER USING ELSE statements]( https://dev.to/dglsparsons/write-better-code-and-be-a-better-programmer-by-never-using-else-statements-4dbl)

You may not feel much difference here, but the habit of using `else ` in code with high lines has some problems.

**Another way could be as follows:**

```php
    public function loginRedirect(): string
    {
        if (session('magicLogin')) {
            return setting('Auth.redirects')['set_password'];
        }

        $url = setting('Auth.redirects')['login'];

        return $this->getUrl($url);
    }
```
